generate-offline-map-using-android-jetpack-workmanager/src/main/java/com/esri/arcgismaps/sample/generateofflinemapusingandroidjetpackworkmanager/OfflineJobWorker.kt

```diff
@@ -0,0 +1,147 @@
+            // return failure if the created job is null
+                ?: return Result.failure()
+
+        return try {
+            // set this worker to run as a long-running foreground service
+            // this will throw an exception, if the worker is launched when the app
+            // is not in foreground
+            setForeground(createForegroundInfo(0))
+            // check and delete if the offline map package file already exists
+            // this check is needed, if the download has failed midway and is restarted later
+            // by WorkManager
+            File(generateOfflineMapJob.downloadDirectoryPath).deleteRecursively()
+
+            // start the generateOfflineMapJob
+            // this job internally runs on a Dispatchers.IO context, hence this CoroutineWorker
+            // can be run on the default Dispatchers.Default context
+            generateOfflineMapJob.start()
+
+            // collect job progress, wait for the job to finish and get the result
+            val jobResult = coroutineScope {
+                // launch the progress collector in a new coroutine
+                val progressCollectorJob = launch {
+                    // collect on progress until the job has completed in a success/failure
+                    generateOfflineMapJob.progress.takeWhile {
+                        generateOfflineMapJob.status.value != JobStatus.Failed
+                            && generateOfflineMapJob.status.value != JobStatus.Succeeded
+                    }.collect { progress ->
+                        // update the worker progress
+                        setProgress(workDataOf("Progress" to progress))
+                        // update the ongoing progress notification
+                        setForeground(createForegroundInfo(progress))
+                    }
+                }
+                // suspends until the generateOfflineMapJob has completed
+                val result = generateOfflineMapJob.result()
+                // cancel the progress collection coroutine if it is still running
+                progressCollectorJob.cancel()
+                // return the result
+                result
+            }
+            // handle and return the result
+            if (jobResult.isSuccess) {
+                // if the job is successful show a final status notification
+                workerNotification.showStatusNotification("Completed")
+                Result.success()
+            } else {
+                // if the job has failed show a final status notification
+                workerNotification.showStatusNotification("Failed")
+                Result.failure()
+            }
+        } catch (exception: Exception) {
+            // capture and log if exception occurs
+            Log.e(TAG, "Offline map job failed:", exception)
+            // a CancellationException is raised if the work is cancelled manually by the user
+            if (exception !is CancellationException) {
```

Comment:
If it's a `CancellationException`, you need to rethrow it, otherwise you break structured concurrency.

Reply:
In this case, since the `CoroutineWorker.doWork` is the parent/root is re-throwing necessary? I saw [doc](https://developer.android.com/guide/background/persistent/threading/coroutineworker) says as follows which I interpreted as not having to rethrow?  - 

`CoroutineWorker handles stoppages automatically by cancelling the coroutine and propagating the cancellation signals. You don't need to do anything special to handle [work stoppages](https://developer.android.com/topic/libraries/architecture/workmanager/how-to/managing-work#cancelling).`

Reply:
@kaushikrw - I think this is different from what that doc is talking about. If you swallow the CancellationException, code that may run after this suspend function would be executed, even if the coroutine was cancelled. This would break the behaviour of structured concurrency. A longer discussion of this issue [here](https://github.com/Kotlin/kotlinx.coroutines/issues/1814), related to the use of `runCatching`.

Reply:
I think that make sense, since I'm not doing anything special to handle a`CancellationException` and the WorkManager ignores the returned result in case of cancellation. Modified it to rethrow if it's a `CancellationException`.

But the WorkManager [samples](https://github.com/googlecodelabs/android-workmanager/blob/advanced/app/src/main/java/com/example/background/workers/BlurWorker.kt) from Google all seem to be actually swallowing CancellationExceptions. I wonder if it's due to an inherent characteristic of WorkManager cancellation mechanism or an oversight in their samples.

Reply:
Interesting. I would assume it's an oversight in their samples. It was also metioned in the [linked issue](https://github.com/Kotlin/kotlinx.coroutines/issues/1814) that many samples ignore the fact that CancellationException should not be swallowed. It's possible that it is even ok in the case of `CoroutineWorker.doWork`, as long as the calling code does not run any other suspend functions on the coroutine that calls `doWork`. But from my POV that would be an assumption we should not make in our code.